Add ytdl search

cycle-five opened the pull request:
Adds to YoutubeDl the ability to just return the search results as text.

cycle-five committed:
Add ytdl search

cycle-five commented:
What are your linter settings? Mine seem to be conflicting.

GnomedDev commented:
You need to run `cargo +nightly fmt`.

cycle-five committed:
fix fmt

FelixMcFelix commented on src/input/compose.rs:
```diff
@@ -37,4 +37,11 @@ pub trait Compose: Send {
+    /// Uses a query to search for a source and return the results.
+    ///
+    /// This method allows for searching for a source and being able to select from the results.
+    async fn search(&mut self) -> Result<Vec<String>, AudioStreamError> {
```
I don't think that this should be added to the `Compose` trait -- this is not common functionality to all or most Input sources.

cycle-five replied:
removed

FelixMcFelix commented on src/input/sources/ytdl.rs:
```diff
@@ -143,6 +189,10 @@ impl Compose for YoutubeDl {
         })
     }
+
+    async fn search(&mut self) -> Result<Vec<String>, AudioStreamError> {
```
See above re: `Compose::search`.

cycle-five replied:
removed

FelixMcFelix commented on src/input/sources/ytdl.rs:
```diff
@@ -56,6 +56,52 @@ impl YoutubeDl {
         }
     }
 
+    /// Creates a lazy(?) request to search youtube for a videos matching `query`.
```
```suggestion
    /// Creates a lazy request to search Youtube for the first video matching `query`.
```

cycle-five replied:
fixed

FelixMcFelix commented on src/input/sources/ytdl.rs:
```diff
@@ -56,6 +56,52 @@ impl YoutubeDl {
+    /// [`new`]: Self::new
+    #[must_use]
+    pub fn new_yt_search(client: Client, query: String) -> Self {
+        Self::new_ytdl_like(YOUTUBE_DL_COMMAND, client, format!("ytsearch5:{}", query))
```
A) Why 5 specifically? Can this be made user-configurable? If you play this `Compose`, I'd expect it would just play the first so maybe `1` makes for a sensible default unless you want to use search query.

B) An `Input` created from this `Compose` will error out, as you will get 5-lines of newline delimited JSON from yt-dlp. You need to alter the JSON decode to account for this.

cycle-five replied:
I like 5 and it seems reasonable number of search results for a discord bot result. I also made it configurable.

FelixMcFelix commented on src/input/sources/ytdl.rs:
```diff
@@ -56,6 +56,52 @@ impl YoutubeDl {
+                })
+            })?;
+
+        let lines = output
+            .stdout
+            .split(|&b| b == b'\n')
+            .map(|line| format!("{}", String::from_utf8_lossy(line)))
+            .collect::<Vec<_>>();
```
This isn't returning a list of valid URLs:
> `~play cloudkicker`

```rust
Ok([
"Cloudkicker - \"The Discovery\" [Full Album - Official - Correct Track Listing]",
"WVM603q6b1w",
"39:53",
"Best of Cloudkicker",
"_uxmzStcbss",
"3:27:04",
"Cloudkicker - Solitude [Album] (2020)",
"Ej1X5M0MAM8",
"48:32",
 // ...
])
```
This goes against the doc-comment of the function.

cycle-five replied:
I will change the documentation, this is the intended output. I am open to a different format if it's preferred, but this is the information I am pulling purposefully for the feature in my bot I made this change for.

FelixMcFelix replied:
I think if you need all this, then returning `Vec<AuxMetadata>` is probably the right call for exposing all this in a structured way.

cycle-five replied:
done.

FelixMcFelix commented on src/input/sources/ytdl.rs:
```diff
@@ -56,6 +56,52 @@ impl YoutubeDl {
+            format!("ytsearch5:{}", self.url)
+        };
+
+        let ytdl_args = [
+            "-s",
+            &search_str,
+            "--get-id",
+            "--get-title",
+            "--get-duration",
+        ];
+
+        let output = Command::new(self.program)
+            .args(ytdl_args)
+            .output()
+            .await
+            .map_err(|e| {
+                AudioStreamError::Fail(if e.kind() == ErrorKind::NotFound {
+                    format!("could not find executable '{}' on path", self.program).into()
+                } else {
+                    Box::new(e)
+                })
+            })?;
```
While converting `query()` to support ndjson (`self.query_many(1).pop()`?), you could probably replace this function to:
* call an inner query function (`YoutubeDl::query_many(&mut self, n: usize)`?) to get a `Vec<AuxMetadata>`.
* You can then either extract `source_url` if you want a `Vec<String>` or just return all the metadata verbatim.

cycle-five replied:
I just returned it as AuxMetadata, lmk if the approach needs further change,

FelixMcFelix reviewed (dismissed):
I think that this feature is needed and useful, but there's work needing done to make it fit and function as expected if you're able to.

cycle-five commented:
I should have commented originally with the impetus for this feature. I am currently using it in my bot, albeit it is not fully implemented, however the entire purpose is to do a search and return options which the client can then choose from, before it plays the song.

cycle-five commented:
> I think that this feature is needed and useful, but there's work needing done to make it fit and function as expected if you're able to.

I am more than happy to work on this to have it fit properly into songbird, it is absolutely needed for a feature I'm using in my bot, so as long as I can make that work, I'm happy to make whatever changes you feel are appropriate.

FelixMcFelix commented:
Thanks, I'm just keen that it works great for everyone. The main assumption with `Compose` is that you should be able to 'just play it', ideally. I.e., `new_yt_search` should produce an object that will play a search result, or produce a different struct entirely (which does not support `Compose`). I hope that makes sense?

cycle-five committed:
Remove compose, add tests, return AuxMetadata

cycle-five committed:
fix parsing of AuxMetadata and better test

cycle-five commented:
I believe I've made all the requested changes and have got it working with these changes in my bot (which of course is the most important piece in this equation).

I feel that my approach to using the yt-dlp search feature and parsing that into the AuxMetadata is either exactly as it is intended, or completely wrong. 

Also for the future I believe yt-dlp may support other searches, and also are there plans to not need yt-dlp?

FelixMcFelix commented:
Thanks for making changes, I'll take a proper look in the next few days.

> I feel that my approach to using the yt-dlp search feature and parsing that into the AuxMetadata is either exactly as it is intended, or completely wrong.

It looks valid at a glance, thank you. If there are any changes then I'll either push to your branch or provide a patchset.

> Also for the future I believe yt-dlp may support other searches, 

We can support these as and when they come up.

> and also are there plans to not need yt-dlp?

You don't need yt-dlp for songbird, but practically speaking you need it to get any usable data from youtube given the way it works.

FelixMcFelix committed:
Fix playability of `YoutubeDl::new_search`

Refactors such that parsing of (ND)JSON is handled in only one location
now, which allows us to greatly simplify the actual `search` method. The
main change is that any `new_search` is now instantly playable.

FelixMcFelix commented:
@cycle-five please have a look and see that you're okay with the changes I've made here:

* a search query is now playable as users would expect.
* search now calls into the same yt-dlp invocation and parsing logic as query.
* tests have been altered: relying on a specific test result's metadata is a bit flaky. Equally, we really want to make sure we're using sources which are creative-commons or similarly licensed.

cycle-five commented:
Awesome! I'm just going to pull this update and make sure my bot can integrate apropos.

cycle-five commented:
This works exactly as I need, so it has my OK.
